torch/_inductor/comms.py

```diff
@@ -0,0 +1,277 @@
+def tuple_sorted(x):
+    if len(x) == 0:
+        return []
+
+    def sort_func(elem):
+        if isinstance(elem, str):
+            return elem
+        else:
+            # We expect `elem` to be `scheduler.BaseSchedulerNode` type here,
+            # but we are not able to do isinstance assert because of circular dependency
+            return elem.get_name()
+
+    return sorted(x, key=sort_func)
+
+
+def sink_waits(result: List["scheduler.BaseSchedulerNode"]) -> List["scheduler.BaseSchedulerNode"]:
+    """
+    Greedily moves waits as late as possible (i.e. until we reach a use). Optimal in terms of
+    communication overlap.
+    """
+    new_result = []
+    cur_waits = set()
+    for snode in result:
+        if isinstance(snode.node, ir.Wait):
+            cur_waits.add(snode)
+        else:
+            for wait in tuple_sorted(cur_waits):
+                if snode in wait.node_users:
+                    new_result.append(wait)
+                    cur_waits.remove(wait)
+            new_result.append(snode)
+    for snode in tuple_sorted(cur_waits):
+        new_result.append(snode)
+    return new_result
+
+
+def raise_comms(result: List["scheduler.BaseSchedulerNode"]) -> List["scheduler.BaseSchedulerNode"]:
+    """
+    Greedily moves comms as early as possible (i.e. until we reach an input).
+    Optimal in terms of communication overlap.
+    """
+    new_result: List["scheduler.BaseSchedulerNode"] = []
+    cur_comms: List["scheduler.BaseSchedulerNode"] = []
+    for snode in reversed(result):
+        if isinstance(snode.node, ir.CollectiveKernel):
+            cur_comms.append(snode)
+        else:
+            for comm in cur_comms:
+                assert len(comm.args) > 0
+            while len(cur_comms) > 0 and any([snode in comm.args for comm in cur_comms]):
+                comm = cur_comms.pop(0)
+                new_result.append(comm)
+            new_result.append(snode)
+    assert len(cur_comms) <= 1
+    for snode in tuple_sorted(cur_comms):
+        new_result.append(snode)
+    result = new_result[::-1]
+    return result
+
+
+def get_ancestors(node):
+    ancestors = set()
+    cur_nodes = [node]
+    while len(cur_nodes) > 0:
+        new_nodes = []
+        for node in cur_nodes:
+            for inp in node.args:
+                if inp not in ancestors:
+                    ancestors.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return ancestors
+
+
+def get_descendants(node):
+    descendants = set()
+    cur_nodes = [node]
+    while len(cur_nodes) > 0:
+        new_nodes = []
+        for node in cur_nodes:
+            for inp in node.node_users:
+                if inp not in descendants:
+                    descendants.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return descendants
+
+
+def decide_global_ordering_of_comms(nodes: List["scheduler.BaseSchedulerNode"]):
```

Comment:
a couple questions on global ordering

1) is the approach that we are not allowed to change the relative order of comm ops from the original program? 

If so, that's pretty safe, but have you reasoned that it's good enough?  (if we consider TP and DP cases, will we need to move a DP comm past a TP comm?)